cpp/src/arrow/dataset/file_test.cc

```diff
@@ -342,5 +347,118 @@ TEST_F(TestFileSystemDataset, WriteProjected) {
     }
   }
 }
+
+class FileSystemWriteTest : public testing::TestWithParam<std::tuple<bool, bool>> {
+ protected:
+  bool IsParallel() { return std::get<0>(GetParam()); }
+  bool IsSlow() { return std::get<1>(GetParam()); }
+
+  FileSystemWriteTest() { dataset::internal::Initialize(); }
+
+  void TestDatasetWriteRoundTrip(
+      std::function<Result<std::shared_ptr<cp::ExecPlan>>(
+          const cp::BatchesWithSchema& source_data, const FileSystemDatasetWriteOptions&,
+          std::function<Future<util::optional<cp::ExecBatch>>()>*)>
+          plan_factory,
```

Comment:
Consider taking here a `cp::Declaration` (instead of a plan) factory that accepts the `write_options,sink_gen` arguments and creates the nodes after `source_data`, then make the plan (including `source_data`) within the current function instead of within the test cases' code. This should simplify and avoid some repeating code.

Reply:
Thanks, this is much cleaner.